telemetry-proxy/telemetry_proxy.go

```diff
@@ -5,66 +5,122 @@ import (
 	"fmt"
 	"io"
 	"log"
+	"net/http"
 	"os"
 	"time"
 
+	"github.com/gorilla/websocket"
+
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/credentials/insecure"
 
 	pb "github.com/cornellrocketryteam/Ground-Software/telemetry-proxy/proto-out" // Replace with your proto package path
 
 	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
+	"github.com/influxdata/influxdb-client-go/v2/api"
 	"github.com/influxdata/influxdb-client-go/v2/api/write"
 )
 
+// Define the WebSocket upgrader
+var upgrader = websocket.Upgrader{
+	ReadBufferSize:  1024,
+	WriteBufferSize: 1024,
+	CheckOrigin: func(r *http.Request) bool {
+		return true // Allow all origins for now (you might want to restrict this in production)
+	},
+}
+
 func main() {
 	// Set up a connection to the influxdb instance.
 	token := os.Getenv("INFLUXDB_TOKEN")
-	influx_url := "http://"+os.Getenv("INFLUXDB_HOSTNAME")+":8086"
+	influx_url := "http://" + os.Getenv("INFLUXDB_HOSTNAME") + ":8086"
 	influx_client := influxdb2.NewClient(influx_url, token)
 
 	org := "crt"
 	bucket := "telemetry"
 	writeAPI := influx_client.WriteAPIBlocking(org, bucket)
 
 	// Set up a connection to the grpc server
-	address := os.Getenv("FILL_HOSTNAME")+":50051" // Replace with your server address
-	conn, err := grpc.NewClient(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	address := os.Getenv("FILL_HOSTNAME") + ":50051"
+	conn, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
 		log.Fatalf("did not connect: %v", err)
 	}
 	defer conn.Close()
 	c := pb.NewTelemeterClient(conn)
 
-	// Contact the server and print out its response.
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	// Start gRPC stream in a separate Goroutine
+	go receiveTelemetry(c, writeAPI)
 
-	stream, err := c.StreamTelemetry(ctx, &pb.TelemetryRequest{})
-	if err != nil {
-		log.Fatalf("could not receive stream: %v", err)
-	}
+	// Start the WebSocket Server
+	http.HandleFunc("/ws", websocketHandler)
+	log.Fatal(http.ListenAndServe(":8080", nil)) // Listen on port 8080 for WebSocket connections
+
+	// Keep the main thread running to prevent the program from exiting
+	select {}
```

Comment:
Can you just move the receiveTelemetry function into the main thread (replacing the `select {}`) so that we don't need an infinite loop in the main thread? Or is there a reason you start it as a goroutine?